Data/Semiring/Tropical.hs

```diff
@@ -0,0 +1,253 @@
+{-# LANGUAGE DeriveFoldable      #-}
+{-# LANGUAGE DeriveFunctor       #-}
+{-# LANGUAGE DeriveTraversable   #-}
+{-# LANGUAGE GADTSyntax          #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving  #-}
+
+-----------------------------------------------------------------------------
+-- |
+--   A tropical semiring is an extension of another totally ordered
+--   semiring with the operations of minimum or maximum as addition.
+--   The extended semiring is given positive or negative infinity as
+--   its 'zero' element, so that the following holds:
+--
+--   @
+--     'plus' 'Infinity' y = y
+--     'plus' x 'Infinity' = x
+--   @
+--
+--
+--   i.e., In the max-plus tropical semiring (where 'plus' is 'max'),
+--   'Infinity' unifies with the typical interpretation of negative infinity,
+--   and thus it is the identity for the maximum, and in the min-plus tropical
+--   semiring (where 'plus' is 'min'), 'Infinity' unifies with the typical
+--   interpretation of positive infinity, and thus it is the identity for the minimum.
+--
+-----------------------------------------------------------------------------
+
+module Data.Semiring.Tropical
+  ( Tropical(Infinity,Tropical)
+  , Extrema(Minima,Maxima)
+  , Extremum(extremum)
+  ) where
+
+import Control.Applicative
+  ( Alternative(empty,(<|>))
+#if !MIN_VERSION_base(4,8,0)
+  , Applicative(..)
+#endif
+
+#if MIN_VERSION_base(4,8,0)
+  , liftA2
+#endif
+  )
+import Control.Monad (MonadPlus(mzero,mplus))
+#if MIN_VERSION_base(4,9,0)
+import Control.Monad.Fail (MonadFail(fail))
+#endif
+import Control.Monad.Fix (MonadFix(mfix))
+#if MIN_VERSION_base(4,8,0)
+import Control.Monad.Zip (MonadZip(mzipWith))
+#endif
+#if MIN_VERSION_base(4,7,0)
+import Data.Data (Data)
+#endif
+#if !MIN_VERSION_base(4,8,0)
+import Data.Foldable (Foldable)
+import Data.Traversable (Traversable)
+#endif
+#if MIN_VERSION_base(4,6,0)
+import Data.Proxy (Proxy(Proxy))
+#endif
+import Data.Semiring (Semiring(..))
+#if MIN_VERSION_base(4,7,0)
+import Data.Typeable (Typeable)
+#endif
+
+import qualified Data.Monoid as Monoid
+import qualified GHC.Natural as Natural
+
+#if !MIN_VERSION_base(4,6,0)
+-- | On older GHCs, 'Data.Proxy.Proxy' is not polykinded, so we provide our own proxy type for 'Extrema'.
+data EProxy (e :: Extrema) = EProxy
+#endif
+
+-- | A datatype to be used at the kind-level. Its only
+--   purpose is to decide the ordering for the tropical
+--   semiring in a type-safe way.
+data Extrema = Minima | Maxima
+
+-- | The 'Extremum' typeclass exists for us to match on
+--   the kind-level 'Extrema', so that we can recover
+--   which ordering to use in the `Semiring` instance
+--   for 'Tropical`.
+class Extremum (e :: Extrema) where
+  -- unfortunately this has to take a `Proxy` because
+  -- we don't have visual type applications before GHC 8.0
+#if !MIN_VERSION_base(4,6,0)
+  extremum :: EProxy e -> Extrema
+#else
+  extremum :: Proxy e -> Extrema
```

Comment:
If we need `EProxy` anyways, it looks simpler to use it instead of `Proxy` for newer GHCs as well. This will save us a decent amount of `CPP`.

Reply:
Hmm, that's a good point. Especially since there can be no user-implemented instances for this class. I foresee supporting older GHCs for a while too.